routing/DELEGATED_CONTENT_ROUTING_HTTP.md

```diff
@@ -0,0 +1,98 @@
+- Peer IDs are encoded according [PeerID string representation specification](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#string-representation)
+- Multibase bytes are encoded according to [the Multibase spec](https://github.com/multiformats/multibase), and *should* use Base64.
+- Timestamps are Unix millisecond epoch timestamps
+
+## API
+- `GET /v1/providers/{CID}`
+    - Response
+        
+        ```json
+        {
+            "Providers": [
+                {
+					"Peer": {
+						"ID": "12D3K...",
+						"Multiaddrs": ["/ip4/.../tcp/.../p2p/...", "/ip4/..."]
+					}
+                    "Protocols": [
+                        {
+                            "Codec": 2320,
+                            "Payload": { ... }
+                        }
+                    ]
+                }
+            ]
+        }
+        ```
+        
+    - Default limit: 100 providers
+    - Optional query parameters
+        - `transfer` only return providers who support the passed transfer protocols, expressed as a comma-separated list of [multicodec codes](https://github.com/multiformats/multicodec/blob/master/table.csv) in decimal form such as `2304,2320`
+        - `transport` only return providers whose published multiaddrs explicitly support the passed transport protocols, such as `460,478` (`/quic` and `/tls/ws`)
+        - Servers should treat the multicodec codes used in the `transfer` and `transport` parameters as opaque, and not validate them, for forwards compatibility
+- `GET /v1/providers/hashed/{multihash}`
+    - This is the same as `GET /v1/providers/{CID}`, but takes a hashed CID encoded as a [multihash](https://github.com/multiformats/multihash/)
+- `PUT /v1/providers`
+    - Reframe equivalent: Provide
+    - Body
+        ```json
+        {
+            "Signature": "multibase bytes",
+            "Payload": "multibase bytes"
+        }
+        ```
+	    - `Payload` is a multibase-encoded string containing a JSON object with the following schema:
+		```json
+		{
+                "Keys": ["cid1", "cid2"],
+                "Timestamp": 1234,
+                "AdvisoryTTL": 1234,
+                "Provider": {
+					"Peer": {
+						"ID": "12D3K...",
+						"Multiaddrs": ["/ip4/.../tcp/.../p2p/...", "/ip4/..."]
+					}
+                    "Protocols": [
+                        {
+                            "Codec": 1234,
+                            "Payload": { ... }
+                        }
+                    ]
+                }
+		}
+		```
+        - `Signature` is a multibase-encoded signature of the sha256 hash of the `Payload` field, signed using the private key of the Peer ID specified in the `Payload` JSON. See the [Peer ID](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#keys) specification for the encoding of Peer IDs. Servers must verify the payload using the public key from the Peer ID. If the verification fails, the server must return a 403 status code.
```

Comment:
To be able to verify the signature:
- We need to sign specifically the content, not the hash (security stuff regarding future hash bugs)
- The content must have an order, to be able to verify the signature with the public key.

Reply:
The order is defined by the client since the client provides the base-encoded payload that was signed, so it doesn't matter to the server, it verifies by signing the base-encoded payload.  (This is a similar scheme as JWT, see e.g. https://www.rfc-editor.org/rfc/rfc7519#section-3.1)

Reply:
but that defined order must be known from other peers to be able to check that the actual data correspond with the sign, right?

Reply:
Discussed that offline. the whole bytes payload will be signed. The order will be the byte-order recieved/send on the request.